packages/@glimmer/core/src/managers/util.ts

```diff
@@ -0,0 +1,102 @@
+import { CapturedArguments } from '@glimmer/interfaces';
+import { consume } from '@glimmer/validator';
+import { assert } from '@glimmer/util';
+import { DEBUG } from '@glimmer/env';
+import { Args } from '../interfaces';
+
+function convertToInt(prop: number | string | symbol): number | null {
+  if (typeof prop === 'symbol') return null;
+
+  const num = Number(prop);
+
+  if (isNaN(num)) return null;
+
+  return num % 1 === 0 ? num : null;
+}
+
+export function argsProxyFor(capturedArgs: CapturedArguments): Args {
+  const { named, positional } = capturedArgs;
+
+  const namedHandler: ProxyHandler<{}> = {
+    get(_target, prop) {
+      if (named.has(prop as string)) {
+        const ref = named.get(prop as string);
+        consume(ref.tag);
+
+        return ref.value();
+      }
+    },
+
+    has(_target, prop) {
+      return named.has(prop as string);
+    },
+
+    ownKeys(_target) {
+      return named.names;
+    },
+
+    getOwnPropertyDescriptor(_target, prop) {
+      if (DEBUG) {
+        assert(
+          named.has(prop as string),
+          'args proxies do not have real property descriptors, so you should never need to call getOwnPropertyDescriptor yourself. This code exists for enumerability, such as in for-in loops and Object.keys()'
+        );
+      }
+      return {
+        enumerable: true,
+        configurable: true,
+      };
+    },
+  };
+
+  const positionalHandler: ProxyHandler<[]> = {
+    get(target, prop) {
+      const parsed = convertToInt(prop);
+
+      if (parsed !== null && parsed < positional.length) {
+        const ref = positional.at(parsed);
+        consume(ref.tag);
+
+        return ref.value();
+      } else if (prop === 'length') {
+        consume(positional.tag);
+        return positional.length;
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      return (target as any)[prop];
+    },
+
+    has(_target, prop) {
+      return typeof prop === 'number' && prop < positional.length;
+    },
+  };
+
+  const namedTarget = {};
+  const positionalTarget: unknown[] = [];
+
+  if (DEBUG) {
+    const setHandler = function(_target: unknown, prop: symbol | string | number): never {
+      throw new Error(
+        `You attempted to set ${String(
+          prop
+        )} on the arguments of a component, helper, or modifier. Arguments are immutable and cannot be updated directly, they always represent the values that is passed down. If you want to set default values, you should use a getter and local tracked state instead.`
```

Comment:
When you call `argsProxyFor`, can you pass another argument for the `type`? Ideally we would have a way to figure out the name of the thing too, but that seems good for a future iteration.